### PR TITLE
Allow root sync path to use child sync time

### PR DESCRIPTION
**Pull request by tcrain**
### What changes are proposed in this pull request?

This addresses issue https://github.com/Alluxio/alluxio/issues/15265.

Currently during the metadata sync of a directory, if a child has a more recent sync time than its parent and does not need to be synced then this child path will not be synced, and the sync of the parent will return an error.  As a result the parent's last sync time will not be updated.

This change addresses the issue in two ways.
1. If a child directory does not need to be synced, then it will not be synced. The parent will then update its sync time only to the oldest sync time of a child that was not synced (or the current clock time if all children were synced).
2. If a child file does not need to be synced (but its updated state has already been loaded from the UFS due to the listing of the parent directory) then the sync is still performed (because no additional UFS operations are needed, unless ACL is enabled for the UFS, then an additional UFS call would be needed so the sync is skipped and the time is calculated as in 1.).

### Does this PR introduce any user facing changes?

No

### CODE REVIEW

File: core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java [hunk 1]
```diff
@@ -493,29 +510,38 @@ private void updateMetrics(boolean success, Instant startTime,
...
+  private SyncResult processSyncPath(AlluxioURI path) {
     if (path == null) {
-      return false;
+      return SyncResult.INVALID_RESULT;
     }
+
+    // if we have already loaded the path from the UFS, and the path
+    // is not a directory, then we will always finish the sync
+    // (even if it is not needed) since we already have all the data we need
+    boolean forceSync = mStatusCache.hasStatus(path).map(
```
Comment (tcrain) [inline core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java:521]: This isn't quite true, as with UFS that have ACL enabled we will still need an extra UFS operation to get the ACL. So maybe this option should only be enabled for UFS that only have ACL disabled?

**Comment by tcrain** [comment]
@secfree this incorporates a modified version of your suggestion, let me know what you think

File: core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java (hunk 1 above)
Comment (tcrain) [reply core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java]: Actually I updated this, so that if ACL is enabled and a sync is not needed on a child file then the sync is skipped. Only if ACL is disabled will the sync still happed if it is not needed, because this will not need an extra UFS operation.

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
The code itself mostly LGTM, thanks for the work! I always wanted to add this change but never really got the chance. I have only one question at the moment: 
```
# First I sync the directory
$ alluxio fs ls /dir
# Then I sync a child after 1min
$ alluxio fs ls -f /dir/file
# The child sync time will be updated but not reflected to /dir
# Then I sync the dir again after 5min, and setting the interval to 5min
$ alluxio fs ls -sync.interval=5min /dir
# Now /dir should sync but /dir/file should not, right? And is that good from our perspective?
```

File: core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java [hunk 2]
```diff
@@ -382,10 +391,17 @@ public SyncStatus sync() throws AccessControlException, InvalidPathException {
...
+          if (!result.isResultValid()) {
+            failedSyncPathCount++;
+          } else if (result.wasSyncPerformed()) {
             syncPathCount++;
           } else {
-            failedSyncPathCount++;
+            skippedSyncPathCount++;
+          }
+          if (result.isResultValid() && !result.wasSyncPerformed()) {
+            childOldestSkippedSync = childOldestSkippedSync == null ? result.getLastSyncTime()
```
Comment (elega) [inline core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java:403]: i'm just thinking aloud here.
if say the interval is 100s, the last synced timestamps are:
/a  0
/a/b 10
/a/c 0
/a/d 0
/a/e 0
/a/f 0
the current timestamp is 100 and a sync will trigger with the sync for /a/b skipped. Then the timestamps look like the following:
/a  10
/a/b 10
/a/c 100
/a/d 100
/a/e 100
/a/f 100

and if we do a sync at timestamp 110, a metadata sync for /a will be triggered again, with only /a/b being sycned. After the operation, the timestamp looks like:
/a  110
/a/b 110
/a/c 100
/a/d 100
/a/e 100
/a/f 100

Is my understanding correct?

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 3]
```diff
@@ -407,6 +408,8 @@ public class DefaultFileSystemMaster extends CoreMaster
   /** Used to check pending/running backup from RPCs. */
   private final CallTracker mStateLockCallTracker;
 
+  final Clock mClock;
```
[the remark this answers is not in the document]
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:411]: I believe this is dependency injection so we can use a manually controlled clock in the UT

File: core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java (hunk 2 above)
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java:403]: My understanding is the same @elega so pls let me know if that's incorrect @tcrain :)

Comment (tcrain) [reply core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java:403]: When a sync happens on a directory, only the sync timestamp of the directory itself will be updated (including information if the sync was recursive). Then whenever a path checked it will check its parent sync time.

I think the reason for this (and why in the previous design indicated a failure if a child was more recent) was so that the sync cache would be updated only on the parent directory level and not for every child synced so that you would have less entires in the cache. And actually updating the children times individually would require a much more extensive redesign because the sync paths are not tracked in a way currently where they can track when their children finish (apart from the root sync path). I kept this idea of only updating the root sync directory instead of updating the individual children, but in reality I don't know what is best and would require benchmarks.

Taking this into account there are two choices for how to solve the problem of a child with a recent sync time.
1. What I have done in this PR, i.e. do not sync the child and only update the root sync path to the oldest skipped sync time.
2. Perform the sync on the child anyway and update the root sync path to the new time.

Lets go through your example, in the solution of 1 (note I am assuming every path here is a directory).

if say the interval is 100s, the last synced timestamps are:
/a 0
/a/b 10
/a/c 0
/a/d 0
/a/e 0
/a/f 0
the current timestamp is 100 and a sync will trigger with the sync for /a/b skipped. Then the timestamps look like the following:
/a 10
/a/b 10
/a/c 0
/a/d 0
/a/e 0
/a/f 0

and if we do a sync at timestamp 110, a metadata sync for /a will be triggered again, all children are synced. After the operation, the timestamp looks like (i.e. all paths have a sync time of 110):
/a 110
/a/b 10
/a/c 0
/a/d 0
/a/e 0
/a/f 0

Now in the solution of 2.
if say the interval is 100s, the last synced timestamps are:
/a 0
/a/b 10
/a/c 0
/a/d 0
/a/e 0
/a/f 0
the current timestamp is 100 and a sync will trigger with the sync for all children. Then the timestamps look like the following:
/a 100
/a/b 10
/a/c 0
/a/d 0
/a/e 0
/a/f 0

and if we do a sync at timestamp 110 no sync is needed:
/a 110
/a/b 10
/a/c 0
/a/d 0
/a/e 0
/a/f 0

Here we see 2. works better than 1., but of course you can come up with an example that is the opposite where 1. works better

For example:

if say the interval is 100s, the last synced timestamps are:
/a 0
/a/b 0
/a/c 0
/a/d 0
/a/e 0
/a/f 0
now say at time 90 some of the children are synced individually. Then the timestamps look like the following:
/a 0
/a/b 0
/a/c 90
/a/d 90
/a/e 90
/a/f 90

and if we do a sync at timestamp 100, a sync will only happen on /a/b, and /a will get updated to 90
/a 90
/a/b 0
/a/c 90
/a/d 90
/a/e 90
/a/f 90

Now in the solution of 2.
if say the interval is 100s, the last synced timestamps are:
/a 0
/a/b 0
/a/c 0
/a/d 0
/a/e 0
/a/f 0
now say at time 90 some of the children are synced individually. Then the timestamps look like the following:
/a 0
/a/b 0
/a/c 90
/a/d 90
/a/e 90
/a/f 90

and if we do a sync at timestamp 100, a a sync will happen on all children and /a will be updated to 100.
/a 100
/a/b 0
/a/c 90
/a/d 90
/a/e 90
/a/f 90

So basically what will work best depends on access pattern.

I chose solution 1 for two reasons:
- It is needed for cross cluster sync (there logical times are used so 1 is much more efficient)
- It gives more flexibility (see how only directories are skipped, and children that are files are still synced since the information is already available from listing the parent)

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tcrain) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:411]: Yes it allows for reliable testing. I believe it is also the preferred way to use a clock since java 8. https://docs.oracle.com/javase/8/docs/api/java/time/Clock.html

File: core/server/master/src/main/java/alluxio/master/file/meta/SyncCheck.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+      mPerformedSync = syncPerformed;
+      mResultValid = resultValid;
+      mLastSyncTime = lastSyncTime;
+    }
+
+    /**
+     * @return true if the result was due to a valid sync (eiter a successful sync,
+     * or a sync skipped due to a recent existing sync)
+     */
+    public boolean isResultValid() {
```
[the remark this answers is not in the document]
Comment (tcrain) [reply core/server/master/src/main/java/alluxio/master/file/meta/SyncCheck.java:104]: The "is" part was actually suggested by Intellij. I guess boolean getters are usually set up this way https://stackoverflow.com/questions/3874350/naming-conventions-for-java-methods-that-return-boolean

**Comment by tcrain** [comment]
> The code itself mostly LGTM, thanks for the work! I always wanted to add this change but never really got the chance. I have only one question at the moment:
> 
> ```
> # First I sync the directory
> $ alluxio fs ls /dir
> # Then I sync a child after 1min
> $ alluxio fs ls -f /dir/file
> # The child sync time will be updated but not reflected to /dir
> # Then I sync the dir again after 5min, and setting the interval to 5min
> $ alluxio fs ls -sync.interval=5min /dir
> # Now /dir should sync but /dir/file should not, right? And is that good from our perspective?
> ```

It sort of depends on the case, but you have the idea.
- If /dir/file is a file and ACL is disabled/unsupported then it and /dir will still be synced (because not further UFS accesses are needed because the updated state of /dir/file already exists from the listing of the parent)
- Otherwise if /dir/file is a directory or ACL is enabled then the sync will be skipped on /dir/file, but the other contents of /dir will still be synced.

File: core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java (hunk 2 above)
Comment (elega) [reply core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java:403]: thanks for the explanation i believe solution 1 is okay for now. I'm thinking in the future we can also probably record the time it took to do a metadata sync for a directory. With the help of such data, we can be adaptive between solution 1 or solution 2. For example, if you know that under a directory there's only one sub directory synced early and it only took half a second to finish. It is totally fine to go with solution 2 to sync everything because the cost is really low. On the other hand if we see a sub directory just synced or took a lot of time to sync, we can go with solution 1 to skip it. 

btw do you want to write down the example of your solution 1 as a comment in the code?

**Review by elega — APPROVED** [review]
LGTM mostly. Please go ahead resolving ongoing discussions with @jiacheliu3 to get his stamp.

File: core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java (hunk 2 above)
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java:403]: @tcrain Thanks for the detailed examples! I overlooked the fact that children sync times are not updated but metadata sync checks parent sync times on the traversal. I'm fine with solution 1 and maybe in the future we define some scenarios to heuristically apply solution 2.

**Review by jiacheliu3 — APPROVED** [review]
LGTM, thanks! A minor comment is Yimin's suggestion where you can add the example into code comments.

**Comment by yuzhu** [comment]
alluxio-bot, merge this please

**Comment by secfree** [comment]
> 

@tcrain Thanks for the information and sorry for the late reply.

The PR looks good to me.

I have some concerns with skipping if ACL enabled
1. This will greatly reduce the cases of application.
2. The cost of an extra UFS operation to get the ACL seems ok. In one cluster of my company
    - The time cost of `(listStatus in Alluxio) / (listStatus in HDFS) = 30`
    - After applying #15277, `(listStatus in Alluxio) / (listStatus in HDFS) = 0.33`. (#15277 does not skipping if ACL enabled.)
3. For HDFS, `HdfsUnderFileSystem.listStatus` already fetched fields like `ower`, `group` and `permission` for files under the directory. Is it still needed to do an extra `getAcl()` for each file? Can we skip the `getAcl()` operation if the `cachedStatus` already contains related information?

File: core/server/master/src/main/java/alluxio/master/file/meta/SyncCheck.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+ */
+public class SyncCheck {
+  private final boolean mShouldSync;
+  private final long mLastSyncTime;
+
+  /** Sync is not needed, with no last sync time. **/
+  public static final SyncCheck SHOULD_NOT_SYNC = new SyncCheck(
+      false, 0);
+  /** Sync is needed. **/
+  public static final SyncCheck SHOULD_SYNC = new SyncCheck(
```
Comment (dbw9580) [inline core/server/master/src/main/java/alluxio/master/file/meta/SyncCheck.java:31]: These constants should not be public: they may be mistaken as enum variants and people can be lured into comparing them by doing `if (mySyncCheckObj == SyncCheck.SHOULD_NOT_SYNC)` while the correct way is `if (!mySyncCheckObj.isShouldSync())`. 

I'd suggest to hide these constants behind two public static factory methods `SyncCheck.shouldSync()` and `SyncCheck.shouldNotSync()`.
